File: crm/app/internal/application/conversation_service.go
// Package application — ConversationService: chat ingest + inbox management.
// Pure domain + ports only; no adapter imports.
package application

import (
	"context"
	"fmt"
	"log"

	"github.com/google/uuid"

	"github.com/vantt/data-integration/crm/app/internal/domain"
	"github.com/vantt/data-integration/crm/app/internal/ports"
)

// ConversationService handles chat ingest and inbox operations.
type ConversationService struct {
	convRepo  ports.ConversationRepository
	partyRepo ports.PartyRepository
}

// NewConversationService constructs a ConversationService.
func NewConversationService(
	convRepo ports.ConversationRepository,
	partyRepo ports.PartyRepository,
) *ConversationService {
	return &ConversationService{convRepo: convRepo, partyRepo: partyRepo}
}

// IngestMessage processes one parsed chat message:
//  1. Upsert conversation by (channel, external_thread_id).
//  2. Insert message (idempotent on external_message_id).
//  3. Update last_message_at and unread_count.
//  4. Resolve psid → party: look up party_identity(type='psid', value=sender_psid).
//     If found, set conversation.party_id; otherwise leave null.
func (s *ConversationService) IngestMessage(ctx context.Context, msg domain.ParsedMessage) error {
	if msg.Channel == "" || msg.ExternalThreadID == "" {
		return domain.NewValidationError("channel and external_thread_id are required")
	}
	if msg.ExternalMessageID == "" {
		return domain.NewValidationError("external_message_id is required")
	}
	if msg.SentAt == "" {
		return domain.NewValidationError("sent_at is required")
	}

	now := utcNow()

	// Step 1: upsert conversation (get or create).
	conv := &domain.Conversation{
		ConversationID:   uuid.New().String(),
		Channel:          msg.Channel,
		ExternalThreadID: msg.ExternalThreadID,
		PageID:           msg.PageID,
		Status:           "open",
		UnreadCount:      0,
		CreatedAt:        now,
		UpdatedAt:        now,
	}
	convID, err := s.convRepo.UpsertConversation(ctx, conv)
	if err != nil {
		return fmt.Errorf("conversation service: upsert conversation: %w", err)
	}

	// Step 2: insert message (idempotent).
	message := &domain.Message{
		MessageID:         uuid.New().String(),
		ConversationID:    convID,
		ExternalMessageID: msg.ExternalMessageID,
		Direction:         msg.Direction,
		SenderRef:         msg.SenderRef,
		Body:              msg.Body,
		Attachments:       msg.Attachments,
		SentAt:            msg.SentAt,
	}
	_, inserted, err := s.convRepo.InsertMessage(ctx, message)
	if err != nil {
		return fmt.Errorf("conversation service: insert message: %w", err)
	}
	if !inserted {
		// Duplicate message — nothing more to do.
		return nil
	}

	// Step 3: atomic update of last_message_at (take-max) and unread_count delta.
	// unreadDelta=1 for inbound, 0 for echo/outbound — avoids double-increment on re-ingest
	// because InsertMessage already returned inserted=false (not reached here) for duplicates.
	unreadDelta := 0
	if msg.Direction == "in" {
		unreadDelta = 1
	}
	if err := s.convRepo.UpdateConversationOnMessage(ctx, convID, msg.SentAt, unreadDelta); err != nil {
		return fmt.Errorf("conversation service: update conversation on message: %w", err)
	}

	// Step 4: resolve psid → party (Messenger-specific).
	// Reload to check current party_id; only update if not yet linked.
	existing, err := s.convRepo.GetConversationByID(ctx, convID)
	if err != nil {
		return fmt.Errorf("conversation service: reload conversation for party lookup: %w", err)
	}
	if existing == nil {
		return fmt.Errorf("conversation service: conversation %s not found after upsert", convID)
	}

	// Use sender_ref for psid lookup (customer psid for inbound, page id for echo).
	// For inbound messages, sender_ref == the customer psid.
	psidRef := msg.SenderRef
	if msg.Direction == "in" {
		psidRef = msg.SenderRef
	} else {
		// For echo/outbound, ExternalThreadID IS the customer psid (set by parser).
		psidRef = msg.ExternalThreadID
	}

	if existing.PartyID == nil && psidRef != "" {
		party, lookupErr := s.partyRepo.FindByIdentity(ctx, "psid", psidRef)
		if lookupErr != nil {
			log.Printf("conversation service: psid lookup %s: %v", psidRef, lookupErr)
		} else if party != nil {
			existing.PartyID = &party.PartyID
			// updated_at is owned by trg_conversation_touch trigger.
			if err := s.convRepo.UpdateConversation(ctx, existing); err != nil {
				return fmt.Errorf("conversation service: update conversation party: %w", err)
			}
		}
	}
	return nil
}

// ListInbox returns conversations filtered by optional assignee and/or status.
func (s *ConversationService) ListInbox(ctx context.Context, assigneeUserID, status string) ([]domain.Conversation, error) {
	convs, err := s.convRepo.ListConversations(ctx, assigneeUserID, status)
	if err != nil {
		return nil, fmt.Errorf("conversation service: list inbox: %w", err)
	}
	if convs == nil {
		convs = []domain.Conversation{}
	}
	return convs, nil
}

// AssignConversation assigns a conversation to a staff user.
func (s *ConversationService) AssignConversation(ctx context.Context, conversationID, assigneeUserID string) error {
	conv, err := s.convRepo.GetConversationByID(ctx, conversationID)
	if err != nil {
		return fmt.Errorf("conversation service: assign: %w", err)
	}
	if conv == nil {
		return domain.NewValidationError(fmt.Sprintf("conversation %q not found", conversationID))
	}
	conv.AssigneeUserID = &assigneeUserID
	// updated_at is owned by trg_conversation_touch trigger — no app-side stamp needed.
	return s.convRepo.UpdateConversation(ctx, conv)
}

// SetStatus transitions a conversation's status (open|pending|closed).
func (s *ConversationService) SetStatus(ctx context.Context, conversationID, newStatus string) error {
	if !domain.IsValidConversationStatus(newStatus) {
		return domain.NewValidationError(fmt.Sprintf("invalid conversation status %q", newStatus))
	}
	conv, err := s.convRepo.GetConversationByID(ctx, conversationID)
	if err != nil {
		return fmt.Errorf("conversation service: set status: %w", err)
	}
	if conv == nil {
		return domain.NewValidationError(fmt.Sprintf("conversation %q not found", conversationID))
	}
	if err := domain.CanTransitionConvTo(conv.Status, newStatus); err != nil {
		return domain.NewValidationError(err.Error())
	}
	conv.Status = newStatus
	// updated_at is owned by trg_conversation_touch trigger — no app-side stamp needed.
	return s.convRepo.UpdateConversation(ctx, conv)
}

// GetMessages returns all messages in a conversation ordered by sent_at ASC.
func (s *ConversationService) GetMessages(ctx context.Context, conversationID string) ([]domain.Message, error) {
	msgs, err := s.convRepo.ListMessages(ctx, conversationID)
	if err != nil {
		return nil, fmt.Errorf("conversation service: get messages: %w", err)
	}
	if msgs == nil {
		msgs = []domain.Message{}
	}
	return msgs, nil
}

File: crm/src/application/conversation_service.py
"""Application service: chat ingest and inbox management.

Pure domain + ports only; no adapter imports.
"""
from __future__ import annotations

import logging
import uuid
from shared.timestamps import utc_now
from typing import Optional

from domain.entities.conversation import (
    Conversation,
    Message,
    ParsedMessage,
    CONV_ALLOWED_TRANSITIONS,
    VALID_CONVERSATION_STATUSES,
    DIRECTION_IN,
)
from domain.entities.party import Party
from domain.ports.conversation_repository import ConversationRepository
from domain.ports.party_repository import PartyRepository

log = logging.getLogger(__name__)


class ConversationService:
    """Handles chat ingest and inbox operations."""

    def __init__(
        self,
        conv_repo: ConversationRepository,
        party_repo: PartyRepository,
    ) -> None:
        self._conv_repo = conv_repo
        self._party_repo = party_repo

    # ------------------------------------------------------------------
    # Ingest
    # ------------------------------------------------------------------

    def ingest_message(self, msg: ParsedMessage) -> None:
        """Process one parsed chat message.

        1. Upsert conversation by (channel, external_thread_id).
        2. Insert message (idempotent on external_message_id).
        3. Update last_message_at (take-max) and unread_count delta.
        4. Resolve psid → party and link if not yet linked.
        """
        if not msg.channel or not msg.external_thread_id:
            raise ValueError("channel and external_thread_id are required")
        if not msg.external_message_id:
            raise ValueError("external_message_id is required")
        if not msg.sent_at:
            raise ValueError("sent_at is required")

        now = utc_now()

        # Step 1: upsert conversation.
        conv = Conversation(
            conversation_id=str(uuid.uuid4()),
            channel=msg.channel,
            external_thread_id=msg.external_thread_id,
            page_id=msg.page_id,
            status="open",
            unread_count=0,
            created_at=now,
            updated_at=now,
        )
        conv_id = self._conv_repo.upsert_conversation(conv)

        # Step 2: insert message (idempotent).
        message = Message(
            message_id=str(uuid.uuid4()),
            conversation_id=conv_id,
            external_message_id=msg.external_message_id,
            direction=msg.direction,
            sender_ref=msg.sender_ref,
            body=msg.body,
            attachments=msg.attachments,
            sent_at=msg.sent_at,
            created_at=now,
        )
        _, inserted = self._conv_repo.insert_message(message)
        if not inserted:
            return  # duplicate — nothing more to do

        # Step 3: advance last_message_at and increment unread_count.
        # unread_delta=1 for inbound only; duplicates are already gated above.
        unread_delta = 1 if msg.direction == DIRECTION_IN else 0
        self._conv_repo.update_conversation_on_message(conv_id, msg.sent_at, unread_delta)

        # Step 4: resolve psid → party (Messenger-specific; harmless for others).
        existing = self._conv_repo.get_conversation_by_id(conv_id)
        if existing is None:
            raise RuntimeError(f"conversation {conv_id!r} not found after upsert")

        # For inbound: customer psid is sender_ref.
        # For echo/outbound: customer psid is external_thread_id (keyed on customer).
        psid_ref = msg.sender_ref if msg.direction == DIRECTION_IN else msg.external_thread_id

        if existing.party_id is None and psid_ref:
            try:
                party = self._party_repo.find_by_identity("psid", psid_ref)
            except Exception as exc:
                log.warning("psid lookup %s: %s", psid_ref, exc)
                party = None
            if party is not None:
                existing.party_id = party.party_id
                self._conv_repo.update_conversation(existing)

    # ------------------------------------------------------------------
    # Inbox
    # ------------------------------------------------------------------

    def list_inbox(
        self,
        assignee_user_id: str = "",
        status: str = "",
    ) -> list[Conversation]:
        """Return conversations filtered by optional assignee and/or status."""
        convs = self._conv_repo.list_conversations(assignee_user_id, status)
        return convs if convs is not None else []

    def list_messages(self, conversation_id: str, limit: int = 100) -> list[Message]:
        """Return messages for a conversation ordered by sent_at ASC."""
        msgs = self._conv_repo.list_messages(conversation_id)
        return (msgs if msgs is not None else [])[:limit]

    # ------------------------------------------------------------------
    # Mutations
    # ------------------------------------------------------------------

    def assign_conversation(self, conversation_id: str, assignee_user_id: str) -> None:
        """Assign a conversation to a staff user."""
        conv = self._get_or_raise(conversation_id)
        conv.assignee_user_id = assignee_user_id
        self._conv_repo.update_conversation(conv)

    def set_status(self, conversation_id: str, new_status: str) -> None:
        """Transition a conversation status (open|pending|closed)."""
        if new_status not in VALID_CONVERSATION_STATUSES:
            raise ValueError(f"invalid conversation status {new_status!r}")
        conv = self._get_or_raise(conversation_id)
        allowed = CONV_ALLOWED_TRANSITIONS.get(conv.status, [])
        if new_status not in allowed:
            raise ValueError(
                f"cannot transition conversation from {conv.status!r} to {new_status!r}"
            )
        conv.status = new_status
        self._conv_repo.update_conversation(conv)

    def link_party(self, conversation_id: str, party_id: str) -> None:
        """Directly link a conversation to a known party."""
        conv = self._get_or_raise(conversation_id)
        conv.party_id = party_id
        self._conv_repo.update_conversation(conv)

    def resolve_party_from_external_id(
        self, channel: str, sender_ref: str
    ) -> Optional[Party]:
        """Look up a party by channel identity (e.g. psid for Messenger)."""
        identity_type = "psid" if channel == "messenger" else channel
        try:
            return self._party_repo.find_by_identity(identity_type, sender_ref)
        except Exception as exc:
            log.warning("resolve_party_from_external_id %s/%s: %s", channel, sender_ref, exc)
            return None

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _get_or_raise(self, conversation_id: str) -> Conversation:
        conv = self._conv_repo.get_conversation_by_id(conversation_id)
        if conv is None:
            raise ValueError(f"conversation {conversation_id!r} not found")
        return conv
